Spigot-API-Patches/0038-Add-Profile-API.patch

```diff
@@ -1071,6 +1086,363 @@ index 0000000..bdfdba1
++        this.cause = cause;
++        setResult(result);
++        this.mojang = mojang;
++    }
++
++    /**
++     * Return the result of this lookup
++     *
++     * @return the result of the lookup
++     */
++    public PlayerProfile getResult() {
++        return result;
++    }
++
++    /**
++     * Set the result of this lookup
++     * <p>Can't be null. If the lookup is a properties lookup, the properties must be set.</p>
++     *
++     * @param result the result of the lookup
++     */
++    public void setResult(PlayerProfile result) {
++        Preconditions.checkNotNull(result, "Null result");
++        if (this.getCause() == LookupCause.PROPERTIES_LOOKUP) {
++            Preconditions.checkArgument(result.hasProperties(), "Result doesn't have properties in properties lookup: %s", result);
++        }
++        this.result = result;
++        this.mojang = false;
++    }
++
++    /**
++     * Get what caused this lookup
++     *
++     * @return what caused this lookup
++     */
++    public LookupCause getCause() {
++        return cause;
++    }
++
++    /**
++     * Return if the profile is known to come from mojang
++     * <p>If not, it is probably from a plugin.</p>
++     *
++     * @return if the profile is known to come from mojang
++     */
++    public boolean isFromMojang() {
++        return mojang;
++    }
++
++    private static final HandlerList handlerList = new HandlerList();
++
++    public HandlerList getHandlerList() {
++        return handlerList;
++    }
++
++    @Override
++    public HandlerList getHandlers() {
++        return getHandlerList();
++    }
++}
+\ No newline at end of file
+diff --git a/src/main/java/com/destroystokyo/paper/profile/event/AsyncPropertiesPreResolveEvent.java b/src/main/java/com/destroystokyo/paper/profile/event/AsyncPropertiesPreResolveEvent.java
+new file mode 100644
+index 0000000..7d9c79a
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/profile/event/AsyncPropertiesPreResolveEvent.java
+@@ -0,0 +1,68 @@
++package com.destroystokyo.paper.profile.event;
++
++import com.destroystokyo.paper.profile.ProfileProperties;
++import com.google.common.base.Preconditions;
++
++import org.bukkit.event.HandlerList;
++import com.destroystokyo.paper.profile.LookupCause;
++import com.destroystokyo.paper.profile.PlayerProfile;
++
++/**
++ * Called before profile properties are requested from mojang.
++ * Plugins can set a profile, which will effectively 'cancel' the lookup from mojang.
++ * <p>The event may be called synchronously if the uuid is resolved on the main thread</p>
++ */
++public class AsyncPropertiesPreResolveEvent extends AsyncProfilePreResolveEvent {
++    private final PlayerProfile profile;
++
++    public AsyncPropertiesPreResolveEvent(PlayerProfile profile) {
++        super(LookupCause.PROPERTIES_LOOKUP);
++        this.profile = profile.withProperties(null);
++        Preconditions.checkNotNull(profile, "Null profile");
```

Comment:
Is this intended to be here? You're checking the local variable if it's null *after* accessing it.